File: android/app/src/main/java/com/reactnativehelloagainworld/MotionDnaReactBridge.java
package com.reactnativehelloagainworld;

import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Handler;
import android.os.Looper;

import android.util.Log;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.navisens.motiondnaapi.MotionDna;
import com.navisens.motiondnaapi.MotionDnaApplication;
import com.navisens.motiondnaapi.MotionDnaInterface;

import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;

import static com.navisens.motiondnaapi.MotionDna.MotionType.FIDGETING;
import static com.navisens.motiondnaapi.MotionDna.MotionType.FORWARD;
import static com.navisens.motiondnaapi.MotionDna.MotionType.STATIONARY;

/**
 * Created by jamesgrantham on 3/15/18.
 */

public class MotionDnaReactBridge extends ReactContextBaseJavaModule implements MotionDnaInterface {

//    MotionDnaApplication motionDnaApplication;
    MotionDnaSynchronizedApplication motionDnaApplication;

    public MotionDnaReactBridge(ReactApplicationContext reactContext) {
        super(reactContext);
        Looper.prepare();

    }

    @ReactMethod
    public void runMotionDna(final String key, final Callback initializationCallback)
    {
        Handler NavisensHandler = new Handler(Looper.getMainLooper());
        NavisensHandler.post(new Runnable() {
            @Override
            public void run() {
                //        motionDnaApplication = new MotionDnaApplication(this);
                motionDnaApplication = new MotionDnaSynchronizedApplication(MotionDnaReactBridge.this);
                motionDnaApplication.runMotionDna(key);

                Timer timer = new Timer();
                timer.schedule(new TimerTask() {
                    @Override
                    public void run() {
                        initializationCallback.invoke();
                    }
                },1000);
            }
        });

    }

    @ReactMethod
    public void setExternalPositioningState(String state)
    {
        if (state.equals("OFF")) {
            motionDnaApplication.setExternalPositioningState(MotionDna.ExternalPositioningState.OFF);
        }
        else if (state.equals("HIGH_ACCURACY")){
            motionDnaApplication.setExternalPositioningState(MotionDna.ExternalPositioningState.HIGH_ACCURACY);
        }
        else if (state.equals("LOW_ACCURACY")) {
            motionDnaApplication.setExternalPositioningState(MotionDna.ExternalPositioningState.LOW_ACCURACY);
        } else {
            motionDnaApplication.setExternalPositioningState(MotionDna.ExternalPositioningState.EXTERNAL_UNDEFINED);
        }
    }

    @ReactMethod
    public void setPowerMode(String powerMode)
    {
        if (powerMode.equals("SUPER_LOW_CONSUMPTION")) {
            motionDnaApplication.setPowerMode(MotionDna.PowerConsumptionMode.SUPER_LOW_CONSUMPTION);
        }
        else if (powerMode.equals("LOW_CONSUMPTION")){
            motionDnaApplication.setPowerMode(MotionDna.PowerConsumptionMode.LOW_CONSUMPTION);
        }
        else if (powerMode.equals("MEDIUM_CONSUMPTION")) {
            motionDnaApplication.setPowerMode(MotionDna.PowerConsumptionMode.MEDIUM_CONSUMPTION);
        }
        else if (powerMode.equals("PERFORMANCE")) {
            motionDnaApplication.setPowerMode(MotionDna.PowerConsumptionMode.PERFORMANCE);
        }
    }


    @ReactMethod
    public void resume()
    {
        motionDnaApplication.resume();
    }

    @ReactMethod
    public void pause()
    {
        motionDnaApplication.pause();
    }

    @ReactMethod
    public void stop()
    {
        motionDnaApplication.stop();
    }

    @ReactMethod
    public void resetLocalEstimation()
    {
        motionDnaApplication.resetLocalEstimation();
    }

    @ReactMethod
    public void resetLocalHeading()
    {
        motionDnaApplication.resetLocalHeading();
    }

    @ReactMethod
    public void setARMModeEnabled(Boolean mode)
    {
        motionDnaApplication.setARModeEnabled(mode);
    }

    @ReactMethod
    public void setBinaryFileLoggingEnabled(Boolean state)
    {
        motionDnaApplication.setBinaryFileLoggingEnabled(state);
    }

    @ReactMethod
    public void setBackpropagationEnabled(boolean state) {
        motionDnaApplication.setBackpropagationEnabled(state);
    }

    @ReactMethod
    public void setBackpropagationBufferSize(int backPropBufferSize) {
        motionDnaApplication.setBackpropagationBufferSize(backPropBufferSize);
    }

    @ReactMethod
    public void setCallbackUpdateRateInMs(Double rate)
    {
        motionDnaApplication.setCallbackUpdateRateInMs(rate);
    }

    @ReactMethod
    public void setNetworkUpdateRateInMs(Double rate)
    {
        motionDnaApplication.setNetworkUpdateRateInMs(rate);
    }

    @ReactMethod
    public void setPowerMode(MotionDna.PowerConsumptionMode mode)
    {
        motionDnaApplication.setPowerMode(mode);
    }

    @ReactMethod
    public void setLocationNavisens()
    {
        motionDnaApplication.setLocationNavisens();
    }

    @ReactMethod
    public void setLocationGPS()
    {
        motionDnaApplication.setLocationGPSOnly();
    }

    @ReactMethod
    public void setLocationLatitudeLongitude(Double lat, Double lng)
    {
        motionDnaApplication.setLocationLatitudeLongitude(lat,lng);
    }

    @ReactMethod
    public void setLocationLatitudeLongitudeAndHeadingInDegrees(Double lat, Double lng, Double angle)
    {
        motionDnaApplication.setLocationLatitudeLongitudeAndHeadingInDegrees(lat,lng,angle);
    }

    @ReactMethod
    public void setHeadingMagInDegrees()
    {
        motionDnaApplication.setHeadingMagInDegrees();
    }

    @ReactMethod
    public void setHeadingInDegrees(Double heading)
    {
        motionDnaApplication.setHeadingInDegrees(heading);
    }

    @ReactMethod
    public void setLocalHeadingOffsetInDegrees(Double degrees)
    {
        motionDnaApplication.setLocalHeadingOffsetInDegrees(degrees);
    }

    @ReactMethod
    public void startUDP()
    {
        motionDnaApplication.startUDP();
    }

    @ReactMethod
    public void startUDP(String room)
    {
        motionDnaApplication.startUDP(room);
    }

    @ReactMethod
    public void stopUDP()
    {
        motionDnaApplication.stopUDP();
    }

    @ReactMethod
    public void setUDPRoom(String room)
    {
        motionDnaApplication.setUDPRoom(room);
    }

    @ReactMethod
    public void sendUDPPacket(String msg)
    {
        motionDnaApplication.sendUDPPacket(msg);
    }

    @ReactMethod
    public void sendUDPQueryRooms(String[] rooms)
    {
        motionDnaApplication.sendUDPQueryRooms(rooms);
    }

    @ReactMethod
    public void start()
    {
        motionDnaApplication.setExternalPositioningState(MotionDna.ExternalPositioningState.LOW_ACCURACY);
//        motionDnaApplication.setLocationAndHeadingGPSMag();
        motionDnaApplication.setLocationLatitudeLongitudeAndHeadingInDegrees(37.787582, -122.396627, 0);

        motionDnaApplication.setPowerMode(MotionDna.PowerConsumptionMode.PERFORMANCE);
//        motionDnaApplication.startUDPHostAndPort("45.79.101.164", "6512");
        motionDnaApplication.setCallbackUpdateRateInMs(100);

//        motionDnaApplication.setARModeEnabled(true);
//        motionDnaApplication.setARModeEnabled(false);
//
//        motionDnaApplication.setContinuousService(true);
//        motionDnaApplication.setContinuousService(false);
        Log.v(this.getClass().getSimpleName(),"Setup MotionDNAApplication");
        if (MainApplication.context == null) {
            Log.v(this.getClass().getSimpleName(),"Context is null");

        } else {
            Log.v(this.getClass().getSimpleName(),"Context is valid");
        }
    }

    // name used to access bridge from javascript side
    @Override
    public String getName() {
        return "MotionDnaReactBridge";
    }

    private void sendEvent(ReactContext reactContext,
                           String eventName,
                           @Nullable WritableMap params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    @Override
    public void receiveMotionDna(MotionDna motionDna)
    {
        String str = "Lat: " + motionDna.getLocation().globalLocation.latitude + "/Lon: " + motionDna.getLocation().globalLocation.longitude + "\n";
        str +="X:" + motionDna.getLocation().localLocation.x + "/Y:" + motionDna.getLocation().localLocation.y + "/Z:" + motionDna.getLocation().localLocation.z + "\n";
        str += "Hdg: " + motionDna.getLocation().heading +  " \n";
//            str += motionDna.getMotion().primaryMotion + "\n";
        str += MotionDnaApplication.checkSDKVersion();

        WritableMap motionDnaParameters = Arguments.createMap();
        motionDnaParameters.putString("MotionDnaString",str);
        motionDnaParameters.putString("location_locationStatus",StringFromLocationStatus(motionDna.getLocation().locationStatus));
        motionDnaParameters.putDouble("location_localLocation_x",motionDna.getLocation().localLocation.x);
        motionDnaParameters.putDouble("location_localLocation_y",motionDna.getLocation().localLocation.y);
        motionDnaParameters.putDouble("location_localLocation_z",motionDna.getLocation().localLocation.z);
        motionDnaParameters.putDouble("location_globalLocation_latitude",motionDna.getLocation().globalLocation.latitude);
        motionDnaParameters.putDouble("location_globalLocation_longitude",motionDna.getLocation().globalLocation.longitude);
        motionDnaParameters.putDouble("location_globalLocation_altitude",motionDna.getLocation().globalLocation.altitude);
        motionDnaParameters.putDouble("location_heading",motionDna.getLocation().heading);
        motionDnaParameters.putDouble("location_localHeading",motionDna.getLocation().localHeading);
        motionDnaParameters.putDouble("location_uncertainty_x",motionDna.getLocation().uncertainty.x);
        motionDnaParameters.putDouble("location_uncertainty_y",motionDna.getLocation().uncertainty.y);
        motionDnaParameters.putInt("location_verticalMotionStatus",motionDna.getLocation().verticalMotionStatus.ordinal());
        motionDnaParameters.putDouble("location_floor",motionDna.getLocation().floor);
        motionDnaParameters.putInt("GPSLocation_locationStatus",motionDna.getGpsLocation().locationStatus.ordinal());
        motionDnaParameters.putDouble("GPSLocation_localLocation_x",motionDna.getGpsLocation().localLocation.x);
        motionDnaParameters.putDouble("GPSLocation_localLocation_y",motionDna.getGpsLocation().localLocation.y);
        motionDnaParameters.putDouble("GPSLocation_localLocation_z",motionDna.getGpsLocation().localLocation.z);
        motionDnaParameters.putDouble("GPSLocation_globalLocation_latitude",motionDna.getGpsLocation().globalLocation.latitude);
        motionDnaParameters.putDouble("GPSLocation_globalLocation_longitude",motionDna.getGpsLocation().globalLocation.longitude);
        motionDnaParameters.putDouble("GPSLocation_globalLocation_altitude",motionDna.getGpsLocation().globalLocation.altitude);
        motionDnaParameters.putDouble("GPSLocation_heading",motionDna.getGpsLocation().heading);
        motionDnaParameters.putDouble("GPSLocation_localHeading",motionDna.getGpsLocation().localHeading);
        motionDnaParameters.putDouble("GPSLocation_uncertainty_x",motionDna.getGpsLocation().uncertainty.x);
        motionDnaParameters.putDouble("GPSLocation_uncertainty_y",motionDna.getGpsLocation().uncertainty.y);
        motionDnaParameters.putInt("GPSLocation_verticalMotionStatus",motionDna.getGpsLocation().verticalMotionStatus.ordinal());
        motionDnaParameters.putDouble("GPSLocation_floor",motionDna.getGpsLocation().floor);
        motionDnaParameters.putDouble("attitude_roll",motionDna.getAttitude().roll);
        motionDnaParameters.putDouble("attitude_pitch",motionDna.getAttitude().pitch);
        motionDnaParameters.putDouble("attitude_yaw",motionDna.getAttitude().yaw);
        motionDnaParameters.putString("motion_motionType",StringFromMotionType(motionDna.getMotion().motionType));
        motionDnaParameters.putInt("location_verticalMotionStatus",motionDna.getLocation().verticalMotionStatus.ordinal());
        motionDnaParameters.putDouble("motionStatistics_dwelling",motionDna.getMotionStatistics().dwelling);
        motionDnaParameters.putDouble("motionStatistics_walking",motionDna.getMotionStatistics().walking);
        motionDnaParameters.putDouble("motionStatistics_stationary",motionDna.getMotionStatistics().stationary);
        motionDnaParameters.putString("ID",motionDna.getID());
        motionDnaParameters.putString("deviceName",motionDna.getDeviceName());
        motionDnaParameters.putDouble("quaternion_x",motionDna.getOrientationQuaternion().x);
        motionDnaParameters.putDouble("quaternion_y",motionDna.getOrientationQuaternion().y);
        motionDnaParameters.putDouble("quaternion_z",motionDna.getOrientationQuaternion().z);
        motionDnaParameters.putDouble("quaternion_w",motionDna.getOrientationQuaternion().w);
        motionDnaParameters.putDouble("timestamp",motionDna.getTimestamp());
        sendEvent(getReactApplicationContext(),"MotionDnaEvent",motionDnaParameters);

    }

    private String StringFromMotionType(MotionDna.MotionType mt) {
        switch (mt) {
            case STATIONARY:
                return "STATIONARY";
            case FIDGETING:
                return "FIDGETING";
            case FORWARD:
                return "FORWARD";
            default:
                break;
        }
        return "";
    }

    private String StringFromLocationStatus(MotionDna.LocationStatus ls) {
        switch (ls) {
            case NAVISENS_INITIALIZED:
                return "NAVISENS_INITIALIZED";
            case NAVISENS_INITIALIZING:
                return "NAVISENS_INITIALIZING";
            case UNINITIALIZED:
                return "UNINITIALIZED";
            default:
                return "DEFAULT";
        }
//        return "";
    }

    @Override
    public Context getAppContext() {
        return getReactApplicationContext();
    }

    @Override
    public PackageManager getPkgManager() {
        return MainApplication.context.getPackageManager();
    }

    @Override
    public void reportError(MotionDna.ErrorCode errorCode, String s) {
        String errorCodeString = new String();
        switch (errorCode) {
            case ERROR_SENSOR_TIMING: {
                errorCodeString = "ERROR_SENSOR_TIMING";
                break;
            }
            case ERROR_AUTHENTICATION_FAILED: {
                errorCodeString = "ERROR_AUTHENTICATION_FAILED";
                break;
            }
            case ERROR_SENSOR_MISSING: {
                errorCodeString = "ERROR_SENSOR_MISSING";
                break;
            }
            case ERROR_SDK_EXPIRED: {
                errorCodeString = "ERROR_SDK_EXPIRED";
                break;
            }
            case ERROR_WRONG_FLOOR_INPUT: {
                errorCodeString = "ERROR_WRONG_FLOOR_INPUT";
                break;
            }
            case ERROR_PERMISSIONS: {
                errorCodeString = "ERROR_PERMISSIONS";
                break;
            }
        }
        Log.v(getClass().getSimpleName(), "Error: " + errorCodeString + " -- " + s);
        WritableMap motionDnaErrorParameters = Arguments.createMap();
        motionDnaErrorParameters.putString("errorCode", errorCodeString);
        motionDnaErrorParameters.putString("errorString",s);
        sendEvent(getReactApplicationContext(),"MotionDnaErrorEvent",motionDnaErrorParameters);

    }

    @Override
    public void receiveNetworkData(MotionDna motionDna) {
        System.out.println(motionDna.getDeviceName());
    }

    @Override
    public void receiveNetworkData(MotionDna.NetworkCode networkCode, Map<String, ? extends Object> map) {

    }
}

File: ios/MotionDnaReactBridge.m
//
//  MotionDnaManager.m
//  reactNativeHelloworld
//
//  Created by James Grantham on 3/14/18.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "MotionDnaReactBridge.h"

@implementation MotionDnaReactBridge
RCT_EXPORT_MODULE();
static MotionDnaManager *_motionDnaManager;

+ (MotionDnaManager *)sharedMDNAManager {
  return _motionDnaManager;
}

- (instancetype)init {
  if (self = [super init]) {

  }
  return self;
}

- (NSDictionary *)constantsToExport
{
  return @{ @"statusBarAnimationNone" : @(UIStatusBarAnimationNone),
            @"statusBarAnimationFade" : @(UIStatusBarAnimationFade),
            @"statusBarAnimationSlide" : @(UIStatusBarAnimationSlide) };
};

- (NSArray<NSString *> *)supportedEvents
{
  return @[@"MotionDnaEvent", @"MotionDnaErrorEvent"];
}

+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

-(void)receiveMotionDna:(MotionDna*)motionDna {
  NSLog(@"Calling motionDNA");
//  -(Location)getLocation;
//  -(Location)getGPSLocation;
//  -(Attitude)getAttitude;
//  -(Motion)getMotion;
//  -(CalibrationStatus)getCalibrationStatus;
//  -(NSString*)getID;
//  -(NSString*)getDeviceName;
//  -(MotionStatistics)getMotionStatistics;
//  -(MotionStatistics)getPolygonMotionStatistics;
//  -(OrientationQuaternion)getQuaternion;
//  -(XY)getDebugVector;
  Location location = [motionDna getLocation];
  XYZ localLocation = location.localLocation;
  double heading = location.localHeading;
  Motion motion = [motionDna getMotion];
  
  NSString *motionDNAString = [NSString stringWithFormat:@"X:%.2f Y:%.2f Z:%.2f\nHeading:%.2f\nMotion: %@",localLocation.x,localLocation.y,localLocation.z,heading,[self NSStringFromMotionType:motion.motionType]];
//  NSLog(@"MotionDNA DATA\n%@",motionDNAString);
  NSDictionary *motionDnaDictionary = @{@"MotionDnaString": motionDNAString,
                                        @"location_locationStatus":[self NSStringFromLocationStatus:[motionDna getLocation].locationStatus],
                                        @"location_localLocation_x":@([motionDna getLocation].localLocation.x),
                                        @"location_localLocation_y":@([motionDna getLocation].localLocation.y),
                                        @"location_localLocation_z":@([motionDna getLocation].localLocation.z),
                                        @"location_globalLocation_latitude":@([motionDna getLocation].globalLocation.latitude),
                                        @"location_globalLocation_longitude":@([motionDna getLocation].globalLocation.longitude),
                                        @"location_globalLocation_altitude":@([motionDna getLocation].globalLocation.altitude),
                                        @"location_heading":@([motionDna getLocation].heading),
                                        @"location_localHeading":@([motionDna getLocation].localHeading),
                                        @"location_uncertainty_x":@([motionDna getLocation].uncertainty.x),
                                        @"location_uncertainty_y":@([motionDna getLocation].uncertainty.y),
                                        @"location_verticalMotionStatus":@([motionDna getLocation].verticalMotionStatus),
                                        @"location_floor":@([motionDna getLocation].floor),
                                        @"GPSLocation_locationStatus":@([motionDna getGPSLocation].locationStatus),
                                        @"GPSLocation_localLocation_x":@([motionDna getGPSLocation].localLocation.x),
                                        @"GPSLocation_localLocation_y":@([motionDna getGPSLocation].localLocation.y),
                                        @"GPSLocation_localLocation_z":@([motionDna getGPSLocation].localLocation.z),
                                        @"GPSLocation_globalLocation_latitude":@([motionDna getGPSLocation].globalLocation.latitude),
                                        @"GPSLocation_globalLocation_longitude":@([motionDna getGPSLocation].globalLocation.longitude),
                                        @"GPSLocation_globalLocation_altitude":@([motionDna getGPSLocation].globalLocation.altitude),
                                        @"GPSLocation_heading":@([motionDna getGPSLocation].heading),
                                        @"GPSLocation_localHeading":@([motionDna getGPSLocation].localHeading),
                                        @"GPSLocation_uncertainty_x":@([motionDna getGPSLocation].uncertainty.x),
                                        @"GPSLocation_uncertainty_y":@([motionDna getGPSLocation].uncertainty.y),
                                        @"GPSLocation_verticalMotionStatus":@([motionDna getGPSLocation].verticalMotionStatus),
                                        @"GPSLocation_floor":@([motionDna getGPSLocation].floor),
                                        @"attitude_roll":@([motionDna getAttitude].roll),
                                        @"attitude_pitch":@([motionDna getAttitude].pitch),
                                        @"attitude_yaw":@([motionDna getAttitude].yaw),
                                        @"motion_motionType":[self NSStringFromMotionType:[motionDna getMotion].motionType],
                                        @"calibrationStatus":@([motionDna getCalibrationStatus]),
                                        @"motionStatistics_dwelling":@([motionDna getMotionStatistics].dwelling),
                                        @"motionStatistics_walking":@([motionDna getMotionStatistics].walking),
                                        @"motionStatistics_stationary":@([motionDna getMotionStatistics].stationary),
                                        @"ID":[motionDna getID],
                                        @"deviceName":[motionDna getDeviceName],
                                        @"quaternion_x":@([motionDna getQuaternion].x),
                                        @"quaternion_y":@([motionDna getQuaternion].y),
                                        @"quaternion_z":@([motionDna getQuaternion].z),
                                        @"quaternion_w":@([motionDna getQuaternion].w),
                                        @"timestamp":@([motionDna getTimestamp])

                                        };
  [self sendEventWithName:@"MotionDnaEvent" body:motionDnaDictionary];

}

- (void)reportError:(ErrorCode)error WithMessage:(NSString *)message {
  NSString *errorCodeString;
  switch (error) {
    case SENSOR_TIMING: {
      errorCodeString = @"ERROR_SENSOR_TIMING";
      break;
    }
    case AUTHENTICATION_FAILED: {
      errorCodeString = @"ERROR_AUTHENTICATION_FAILED";
      break;
    }
    case SENSOR_MISSING: {
      errorCodeString = @"ERROR_SENSOR_MISSING";
      break;
    }
    case SDK_EXPIRED: {
      errorCodeString = @"ERROR_SDK_EXPIRED";
      break;
    }
    case WRONG_FLOOR_INPUT: {
      errorCodeString = @"ERROR_WRONG_FLOOR_INPUT";
      break;
    }
  }
  NSDictionary *motionDnaErrorDictionary = @{@"errorCode":errorCodeString,
                                             @"errorString":message};
  [self sendEventWithName:@"MotionDnaErrorEvent" body:motionDnaErrorDictionary];

}


- (NSString *)NSStringFromMotionType:(MotionType)mt {
  switch (mt) {
    case STATIONARY:
      return @"STATIONARY";
      break;
    case FIDGETING:
      return @"FIDGETING";
      break;
    case FORWARD:
      return @"FORWARD";
    default:
      break;
  }
  return nil;
}

- (NSString *)NSStringFromLocationStatus:(LocationStatus)ls {
  switch (ls) {
    case NAVISENS_INITIALIZED:
      return @"NAVISENS_INITIALIZED";
    case NAVISENS_INITIALIZING:
      return @"NAVISENS_INITIALIZING";
    case UNINITIALIZED:
      return @"UNINITIALIZED";
    default:
      break;
  }
  return nil;
}

RCT_EXPORT_METHOD(runMotionDna:(NSString*)key callback:(RCTResponseSenderBlock)initializationCallback)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    if (_motionDnaManager == nil) {
      _motionDnaManager = [[MotionDnaManager alloc] init];
    }
    [_motionDnaManager runMotionDna:key];
    _motionDnaManager.controller = self;
    initializationCallback(@[[NSNull null]]);
  });
}

RCT_EXPORT_METHOD(receiveNetworkData:(MotionDna*)motionDna)
{
  [_motionDnaManager receiveNetworkData:motionDna];
}
//-(void)receiveNetworkData:(NetworkCode)opcode WithPayload:(NSDictionary*)payload;

RCT_EXPORT_METHOD(runMotionDnaWithoutMotionManager:(NSString*)ID)
{
  [_motionDnaManager runMotionDnaWithoutMotionManager:ID];
}

//-(void)reportError:(ErrorCode)error WithMessage:(NSString*)message;
//RCT_EXPORT_METHOD(receiveDeviceMotion:(CMDeviceMotion*)deviceMotion)
//{
//  [_motionDnaManager receiveDeviceMotion:deviceMotion];
//}

RCT_EXPORT_METHOD(setFloorNumber:(int)floor)
{
  [_motionDnaManager setFloorNumber:floor];
}

RCT_EXPORT_METHOD(setLocationLatitude:(double)latitude Longitude:(double)longitude AndHeadingInDegrees:(double)heading)
{
  [_motionDnaManager setLocationLatitude:latitude Longitude:longitude AndHeadingInDegrees:heading];
}

RCT_EXPORT_METHOD(setLocationAndHeadingGPSMag)
{
  [_motionDnaManager setLocationAndHeadingGPSMag];
}

RCT_EXPORT_METHOD(setLocationLatitude:(double)latitude Longitude:(double)longitude)
{
  [_motionDnaManager setLocationLatitude:latitude Longitude:longitude];
}

RCT_EXPORT_METHOD(setLocationGPSOnly)
{
  [_motionDnaManager setLocationGPSOnly];
}
RCT_EXPORT_METHOD(setHeadingMagInDegrees)
{
  [_motionDnaManager setHeadingMagInDegrees];
}
RCT_EXPORT_METHOD(setHeadingInDegrees:(double)heading)
{
  [_motionDnaManager setHeadingInDegrees:heading];
}
RCT_EXPORT_METHOD(setLocationBeacon)
{
  [_motionDnaManager setLocationBeacon];
}
RCT_EXPORT_METHOD(setLocationNavisens)
{
  [_motionDnaManager setLocationNavisens];
}

RCT_EXPORT_METHOD(pause)
{
  [_motionDnaManager pause];
}
RCT_EXPORT_METHOD(resume)
{
  [_motionDnaManager resume];
}

RCT_EXPORT_METHOD(startCalibration)
{
  [_motionDnaManager startCalibration];
}

RCT_EXPORT_METHOD(stop)
{
  [_motionDnaManager stop];
}

//+(NSString*)checkSDKVersion;
//-(NSString*)getDeviceID;
//-(NSString*)getDeviceName;

RCT_EXPORT_METHOD(setMapCorrectionEnabled:(BOOL)state)
{
  [_motionDnaManager setMapCorrectionEnabled:state];
}

RCT_EXPORT_METHOD(setCallbackUpdateRateInMs:(double)rate)
{
  printf("Callback update rate: %0.2f\n",rate);
  [_motionDnaManager setCallbackUpdateRateInMs:rate];
}

RCT_EXPORT_METHOD(setNetworkUpdateRateInMs:(double)rate)
{
  [_motionDnaManager setNetworkUpdateRateInMs:rate];
}

RCT_EXPORT_METHOD(setBinaryFileLoggingEnabled:(BOOL)state)
{
  [_motionDnaManager setBinaryFileLoggingEnabled:state];
}

RCT_EXPORT_METHOD(setBackpropagationEnabled:(BOOL)state)
{
  [_motionDnaManager setBackpropagationEnabled:state];
}

RCT_EXPORT_METHOD(setBackpropagationBufferSize:(NSInteger)backPropBufferSize)
{
  [_motionDnaManager setBackpropagationBufferSize:backPropBufferSize];
}

RCT_EXPORT_METHOD(setExternalPositioningState:(NSString*)state)
{
//  OFF,
//  HIGH_ACCURACY,
//  LOW_ACCURACY,
//  EXTERNAL_UNDEFINED
  if ([state isEqualToString:@"OFF"]) {
    [_motionDnaManager setExternalPositioningState:OFF];
    NSLog(@"Entering %@ mode",state);
  }
  if ([state isEqualToString:@"HIGH_ACCURACY"]) {
    [_motionDnaManager setExternalPositioningState:HIGH_ACCURACY];
    NSLog(@"Entering %@ mode",state);
  }
  if ([state isEqualToString:@"LOW_ACCURACY"]) {
    [_motionDnaManager setExternalPositioningState:LOW_ACCURACY];
    NSLog(@"Entering %@ mode",state);
  }
  else {
    NSLog(@"UNDEFINED");
    [_motionDnaManager setExternalPositioningState:EXTERNAL_UNDEFINED];
  }
}

RCT_EXPORT_METHOD(startUDPRoom:(NSString*)room AtHost:(NSString*)host AndPort:(NSString*)port)
{
  [_motionDnaManager startUDPRoom:room AtHost:host AndPort:port];
}

RCT_EXPORT_METHOD(startUDPHost:(NSString*)host AndPort:(NSString*)port)
{
  [_motionDnaManager startUDPHost:host AndPort:port];
}

RCT_EXPORT_METHOD(startUDPRoom:(NSString*)room)
{
  [_motionDnaManager startUDPRoom:room];
}

RCT_EXPORT_METHOD(startUDP)
{
  [_motionDnaManager startUDP];
}
                  
RCT_EXPORT_METHOD(stopUDP)
{
  [_motionDnaManager stopUDP];
}

RCT_EXPORT_METHOD(setUDPRoom:(NSString*)room)
{
  [_motionDnaManager setUDPRoom:room];
}

RCT_EXPORT_METHOD(sendUDPPacket:(NSString*)msg)
{
  [_motionDnaManager sendUDPPacket:msg];
}

RCT_EXPORT_METHOD(sendUDPQueryRooms:(NSMutableArray*)rooms)
{
  [_motionDnaManager sendUDPQueryRooms:rooms];
}

RCT_EXPORT_METHOD(setBackgroundModeEnabled:(BOOL)state)
{
  [_motionDnaManager setBackgroundModeEnabled:state];
}

RCT_EXPORT_METHOD(logBeaconWithName:(NSString*)name ID:(NSString*)id RSSI:(int)RSSI andTimestamp:(double)timestamp)
{
  [_motionDnaManager logBeaconWithName:name ID:id RSSI:RSSI andTimestamp:timestamp];
}

RCT_EXPORT_METHOD(setBeaconRangingEnabled:(BOOL)state)
{
  [_motionDnaManager setBeaconRangingEnabled:state];
}

RCT_EXPORT_METHOD(setBeaconCorrectionsEnabled:(BOOL)state)
{
  [_motionDnaManager setBeaconCorrectionsEnabled:state];
}

RCT_EXPORT_METHOD(removeBeaconRegionWithUUID:(NSUUID *)proximityUUID andIdentifier:(NSString*)identifierForVendor)
{
  [_motionDnaManager removeBeaconRegionWithUUID:proximityUUID andIdentifier:identifierForVendor];
}

RCT_EXPORT_METHOD(registerBeaconRegionWithUUID:(NSUUID *)proximityUUID andIdentifier:(NSString*)identifier)
{
  [_motionDnaManager registerBeaconRegionWithUUID:proximityUUID andIdentifier:identifier];
}
//-(BOOL)isRangingBLE;

RCT_EXPORT_METHOD(setPowerMode:(NSString *)mode)
{
  if ([mode isEqualToString:@"SUPER_LOW_CONSUMPTION"]) {
    [_motionDnaManager setPowerMode:SUPER_LOW_CONSUMPTION];
    NSLog(@"Entering %@ mode",mode);

  }
  else if ([mode isEqualToString:@"LOW_CONSUMPTION"]) {
    [_motionDnaManager setPowerMode:LOW_CONSUMPTION];
    NSLog(@"Entering %@ mode",mode);
  }
  else if ([mode isEqualToString:@"MEDIUM_CONSUMPTION"]) {
    [_motionDnaManager setPowerMode:MEDIUM_CONSUMPTION];
    NSLog(@"Entering %@ mode",mode);

  }
  else if ([mode isEqualToString:@"PERFORMANCE"]) {
    [_motionDnaManager setPowerMode:PERFORMANCE];
    NSLog(@"Entering %@ mode",mode);

  } else {
    NSLog(@"%@ mode is invalid",mode);
  }
}

RCT_EXPORT_METHOD(setLocalHeadingOffsetInDegrees:(double)hdg)
{
  [_motionDnaManager setLocalHeadingOffsetInDegrees:hdg];
}

RCT_EXPORT_METHOD(setCartesianOffsetInMetersX:(double)x Y:(double)y)
{
  [_motionDnaManager setCartesianOffsetInMetersX:x Y:y];
}

RCT_EXPORT_METHOD(setARModeEnabled:(BOOL)state)
{
  [_motionDnaManager setARModeEnabled:state];
}

RCT_EXPORT_METHOD(setEstimationMode:(EstimationMode)mode)
{
  [_motionDnaManager setEstimationMode:mode];
}

RCT_EXPORT_METHOD(resetLocalEstimation)
{
  [_motionDnaManager resetLocalEstimation];
}

RCT_EXPORT_METHOD(enableBackgroundSensors)
{
  [_motionDnaManager enableBackgroundSensors];
}

RCT_EXPORT_METHOD(start)
{
  //[_motionDnaManager setLocationAndHeadingGPSMag];
  //[_motionDnaManager setLocationNavisens];
  //    [_motionDnaManager setExternalPositioningState:LOW_ACCURACY];
  
//  [_motionDnaManager setBackgroundModeEnabled:YES];
//  [_motionDnaManager setCallbackUpdateRateInMs:100];
//  [_motionDnaManager setNetworkUpdateRateInMs:100];
//  [_motionDnaManager setExternalPositioningState:HIGH_ACCURACY];
}

@end
